### PR TITLE
fix: provide hostname and port to server

### LINKED ISSUE #1141: [Bug]: Next.js middlewares are currently broken on Netlify
### Summary

Next.js throws an error when `hostname` and `port` are not provided in the custom server since v12.0.5-canary.18 — as a result it breaks middleware in Netlify:

[next.js@base-server.ts#L725-L729](https://github.com/vercel/next.js/blob/2c9c8f76a36f919bae8e10d7965463080f10e277/packages/next/server/base-server.ts#L725-L729)
```js
// For middleware to "fetch" we must always provide an absolute URL
const url = getRequestMeta(params.request, '__NEXT_INIT_URL')!
 if (!url.startsWith('http')) {
   throw new Error(
     'To use middleware you must provide a `hostname` and `port` to the Next.js Server'
   )
}
```

Screenshot 2022-01-18 at 23 47 04

Ref: https://github.com/netlify/netlify-plugin-nextjs/blob/main/src/templates/getHandler.ts#L46-L50

### Steps to reproduce

1. Fork this simple Next.js app (`npx create-next-app` with one middleware) https://github.com/ivorpad/repro-nextjs-middleware
2. Deploy to Netlify
3. Go to `/api/users`
4. Watch the `___netlify-handler` function logs

### A link to a reproduction repository

https://github.com/ivorpad/repro-nextjs-middleware

### Plugin version

4.2.0

### More information about your build

- [ ] I am building using the CLI
- [X] I am building using file-based configuration (`netlify.toml`)

### What OS are you using?

_No response_

### Your netlify.toml file

<details>
  <summary>`netlify.toml`</summary>
  
```toml
[build]
command = "npm run build"
functions = "out_functions"
publish = ".next"

[[plugins]]
package = "@netlify/plugin-nextjs"
```

</details>

### Your public/_redirects file

<details>
  <summary>`_redirects`</summary>
  
```toml
# Paste content of your `_redirects` file here
```

</details>

### Your `next.config.js` file

<details>
  <summary>`next.config.js`</summary>
  
```toml
# Paste content of your `next.config.js` file here. Check there is no private info in there.
```

</details>

### Builds logs (or link to your logs)

<details>
  <summary>Build logs</summary>
  
```
# Paste logs here
```

</details>

### Function logs

<details>
  <summary>Function logs</summary>
  
```
# Paste logs here
```

</details>

### .next JSON files

<details>
  <summary>generated .next JSON files</summary>
  
```
# Paste file contents here. Please check there isn't any private info in them
# You can either build locally, or download the deploy from Netlify by clicking the arrow next to the deploy time.
```

</details>

### PR DESCRIPTION
**Pull request by ascorbic**
### Summary

As of `next@12.0.8`, middleware requires `NextServer` to be instantiated with a port and hostname. 

### Test plan

1. Visit middleware page on deploy preview 

### Relevant links (GitHub issues, Notion docs, etc.) or a picture of cute animal

Fixes #1141

capybara

### Standard checks:

- [ ] Check the Deploy Preview's Demo site for your PR's functionality
- [ ] Add docs when necessary

---

🧪 Once merged, make sure to update the version if needed and that it was published correctly.

### CODE REVIEW

File: cypress/integration/default/preview.spec.ts [hunk 1]
```diff
@@ -5,17 +5,15 @@ describe('Preview Mode', () => {
...
+    cy.findByText('Is preview? Yes!', { selector: 'h1' })
 
-  // exit preview mode
-  cy.request('/api/exitPreview')
-  cy.visit('/previewTest')
-  cy.findByText('Is preview? No')
+    // exit preview mode
+    cy.request('/api/exitPreview')
+    cy.visit('/previewTest')
+    cy.findByText('Is preview? No', { selector: 'h1' })
```
Comment (tiffanosaurus) [inline cypress/integration/default/preview.spec.ts:17]: Why is Cypress in CI saying there is more than one instance of this text :confused-capybara:?

File: src/templates/getHandler.ts [hunk 2]
```diff
@@ -41,39 +43,48 @@ const makeHandler =
...
-    })
-    const requestHandler = nextServer.getRequestHandler()
-    const server = new Server(async (req, res) => {
-      try {
-        await requestHandler(req, res)
-      } catch (error) {
-        console.error(error)
-        throw new Error('server function error')
+    let bridge
+    const getBridge = (event: HandlerEvent) => {
```
Comment (tiffanosaurus) [inline src/templates/getHandler.ts:47]:
```suggestion
    const getBridge = (event: HandlerEvent): Bridge => {
```

? I'm guessing it's possibly not an available type from Vercel

File: src/templates/getHandler.ts [hunk 3]
```diff
@@ -41,39 +43,48 @@ const makeHandler =
...
-      customServer: false,
-    })
-    const requestHandler = nextServer.getRequestHandler()
-    const server = new Server(async (req, res) => {
-      try {
-        await requestHandler(req, res)
-      } catch (error) {
-        console.error(error)
-        throw new Error('server function error')
+    let bridge
```
Comment (tiffanosaurus) [inline src/templates/getHandler.ts:46]:
```suggestion
    let bridge: Bridge
```

? I'm guessing it's possibly not an available type from Vercel

File: src/templates/getHandler.ts [hunk 4]
```diff
@@ -41,39 +43,48 @@ const makeHandler =
...
+        hostname: url.hostname,
+        port,
+      })
+      const requestHandler = nextServer.getRequestHandler()
+      const server = new Server(async (req, res) => {
+        try {
+          await requestHandler(req, res)
+        } catch (error) {
+          console.error(error)
+          throw new Error('server function error')
```
Comment (tiffanosaurus) [inline src/templates/getHandler.ts:71]: [sand] Might want the error here to be a little more precise.

**Review by tiffanosaurus — APPROVED** [review]

File: cypress/integration/default/preview.spec.ts (hunk 1 above)
Comment (ascorbic) [reply cypress/integration/default/preview.spec.ts:17]: I was baffled for ages, but if you look at the error in the dashboard, it seems there's a `<p>`, which is the route announcer (for screen readers)! No idea why it only shows for this route though.

File: src/templates/getHandler.ts (hunk 2 above)
Comment (ascorbic) [reply src/templates/getHandler.ts:47]: Yeah, I meant to look at up. Thanks for the reminder. I would say there must be types as they wetite it in TypeScript, but you never can tell with Vercel.

Comment (ascorbic) [reply src/templates/getHandler.ts:47]: Well, whaddayaknow https://github.com/vercel/vercel/blob/main/packages/node-bridge/bridge.d.ts

**Comment by ascorbic** [comment]
I changed the indentation level of the handler, so this is easier to review [with whitespace hidden](https://github.com/netlify/netlify-plugin-nextjs/pull/1149/files?diff=unified&w=1)

**Review by tiffanosaurus — APPROVED** [review]

**Pull request merged**